python/tskit/tables.py

```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
         currently indexed this method has no effect.
         """
         self.ll_tables.drop_index()
+
+    def subset_nodes(self, nodes):
+        # The C version.
+        nodes = util.safe_np_int_cast(nodes, np.int32)
+        self.ll_tables.subset_nodes(nodes)
+
+    def subset(self, nodes, record_provenance=True):
+        """
+        Modifies the tables in place to contain only
+        the entries referring to the provided list of nodes,
+        with nodes reordered according to the order they appear in the list.
+        Specifically, this subsets each of the tables as follows:
+
+        1. Nodes: if in the list of nodes, and in the order provided.
+        2. Individuals and Populations: if referred to by a retained node,
+            and in the order first seen when traversing the list of retained nodes.
+        3. Edges: if both parent and child are retained nodes.
+        4. Mutations: if the mutations' node is a retained node.
+        5. Sites: if any mutations remain at the site after removing mutations.
+        6. Migrations: if the migration's node is a retained node.
+
+        If ``nodes`` is the entire list of nodes in the tables, then the resulting
+        tables will be identical to the original tables, but with nodes
+        (and individuals and populations) reordered.
+
+        **Note:** This is quite different from :meth:`.simplify`: the resulting
+        tables contain only the nodes given, not ancestral ones as well, and
+        does not simplify the relationships in any way.
+
+        :param list[int] nodes: A list of node IDs to retain in the tables.
+        """
+        if nodes.shape[0] == 0:
+            raise ValueError("Nodes cannot be empty.")
+        if np.max(nodes) >= self.nodes.num_rows:
+            raise ValueError("One of the nodes is not in the TableCollection.")
+        tables = self.copy()
+        n = tables.nodes
+        # figuring out which individuals to keep
+        old_inds = n.individual[nodes]
+        indiv_to_keep, i = np.unique(old_inds, return_index=True)
+        # maintaining order in which they appeard
+        indiv_to_keep = indiv_to_keep[np.argsort(i)]
+        # removing -1
+        indiv_to_keep = indiv_to_keep[indiv_to_keep != tskit.NULL]
+        # subsetting individuals table
+        if indiv_to_keep.shape[0] == 0:
```

Comment:
We shouldn't need this special case (esp if we clear the tables above).

Reply:
I think `set_columns` complains about it being empty.